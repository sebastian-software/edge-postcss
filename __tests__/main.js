import postcss from "postcss"
import loadConfig from "postcss-load-config"

let plugins = null
let options = null

beforeAll(() =>
  loadConfig({ map: false }).then((config) => {
    plugins = config.plugins
    options = config.options
    return true
  })
)

// We add some super basic formatting to our CSS to make snapshots better readable
// and inspectable in case of any regressions later on.
function format(cssString) {
  return cssString
    .replace(/;/g, ";\n")
    .replace(/}/g, "\n}\n\n")
    .replace(/{/g, "{\n")
    .replace(/}\n\n\n}/g, "}\n}\n\n")
    .replace(/\n\n\n\n/g, "\n\n")
    .trim()
}

function compile(input) {
  let allOptions = { ...options, from: "__tests__/main.css", to: "__tests__/dist/main.css" }

  return postcss(plugins)
    .process(input, allOptions)
    .then((result) => expect(format(result.css)).toMatchSnapshot())
}


// ====================================================
// ================== MERGING =========================
// ====================================================

test("Smart Import Basic", () =>
  compile("@import './fixtures/import-a.css';")
)

test("Smart Import with Merge", () =>
  compile("@import './fixtures/import-b.css'; .section { background: #333; }")
)



// ====================================================
// ================= URLS/ASSETS ======================
// ====================================================

test("Simple URL", () =>
  compile(".icon { background: url('./fixtures/formula.png'); }")
)

test("Asset Size", () =>
  compile(".icon { background-size: width('./fixtures/formula.png') height('./fixtures/formula.png'); }")
)



// ====================================================
// ================ SASS INSPIRED =====================
// ====================================================

test("Sassy Mixins", () =>
  compile("@mixin simple{ color: red; } h1 { @include simple; }")
)

test("Sassy Variables", () =>
  compile("$bgColor: red; h1 { background: $bgColor; }")
)

test("Nested Basic", () =>
  compile("body { h1 { font-weight: bold; } h2 { font-weight: normal; }}")
)

test("Nested Parent Selector", () =>
  compile("body { ul { li { &:first-child { margin-top: 0; }}}}")
)



// ====================================================
// ================== LAYOUT ==========================
// ====================================================

test("Lost Grid", () =>
  compile(".grid { lost-column: 3/12 }")
)

test("Grid KISS", () =>
  compile(`
    .gridTest {
    	grid-kiss:
    		"+-------------------------------+      "
    		"|           header ↑            | 120px"
    		"+-------------------------------+      "
    		"                                       "
    		"+-- 30% ---+  +--- auto --------+      "
    		"| .sidebar |  |       main      | auto "
    		"+----------+  +-----------------+      "
    		"                                       "
    		"+-------------------------------+      "
    		"|              ↓                | 60px "
    		"|         → footer ←          |      "
    		"+-------------------------------+      "
    }
  `)
)



// ====================================================
// ================== COLOR ===========================
// ====================================================

test("RGBA with Hex", () =>
  compile(".red { color: rgba(#f11, 0.8); }")
)

test("Hex with Alpha", () =>
  compile(".blue { color: #11fc; }")
)

test("Color Function", () =>
  compile(".yellow { color: color(yellow a(90%)) }")
)




// ====================================================
// =============== MEDIA QUERIES ======================
// ====================================================

test("Media Query Min/Max", () =>
  compile("@media screen and (width >= 500px) and (width <= 1200px) { .elem { display: block; } }")
)

test("Custom Media", () =>
  compile(`
    @custom-media --small-viewport (max-width: 30em);
    @media (--small-viewport) {
      body { font-size: 12px; }
    }
  `)
)



// ====================================================
// ================== EFFECTS =========================
// ====================================================

test("Magic Animations", () =>
  compile(".animation { animation-name: magic; }")
)

test("Will Change Compat", () =>
  compile(".scaled { will-change: width; }")
)

test("Easings", () =>
  compile(".snake { transition: all 600ms ease-in-sine; }")
)

test("Pleeease Filters", () =>
  compile(".box { filter: drop-shadow(16px 16px 20px blue); }")
)

test("Transform Shortcut", () =>
  compile(".transform { scale: 2; translate: 10px 20px; }")
)



// ====================================================
// ================= EXTENSIONS =======================
// ====================================================

test("Responsive Type", () =>
  compile("html { font-size: responsive 12px 21px; font-range: 420px 1280px; }")
)

test("Clearfix", () =>
  compile(".row { clear: fix; }")
)

test("System UI", () =>
  compile("body { font-family: system-ui; }")
)



// ====================================================
// ================== FIXES ===========================
// ====================================================

test("Gradient Fix", () =>
  compile(".elem { background-image: linear-gradient(green, transparent); }")
)

test("Flexbox Fix", () =>
  compile(".elem { flex: 1; }")
)

test("Input Style Fixes", () =>
  compile(`
    input[type="range"]::track {
      height: 3px;
    }

    input[type="range"]::thumb {
      width: 16px;
      height: 8px;
    }
  `)
)

test("Autoprefixer", () =>
  compile(":fullscreen a { display: flex }")
)

test("Pseudoelements", () =>
  compile("a::before { color: red; }")
)

test("Selector matches()", () =>
  compile("p:matches(:first-child, .special) { color: red; }")
)



// ====================================================
// ================ OPTIMIZATION ======================
// ====================================================

test("Calc Trivial", () =>
  compile(".elem { width: calc(300px + 10px); }")
)

test("Calc Variable", () =>
  compile("$margin: 10px; .elem { width: calc(200px + $margin); }")
)

test("Calc Keep", () =>
  compile(".elem { width: calc(100px + 2%); }")
)

test("zIndex", () =>
  compile(".first { z-index: 1000; } .second { z-index: 2000; }")
)

test("CSS-O (Optimizer)", () =>
  compile(".elem { color: #ff0000; }")
)
