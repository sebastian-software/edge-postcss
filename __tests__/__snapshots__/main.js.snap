// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Asset Size (Same Folder) 1`] = `
".icon{
background-size:304px 85px
}"
`;

exports[`Asset Size 1`] = `
".icon{
background-size:304px 85px
}"
`;

exports[`Asset URL (Same Folder) 1`] = `
".icon{
background:url(./formula.png)
}"
`;

exports[`Asset URL 1`] = `
".icon{
background:url(../fixtures/formula.png)
}"
`;

exports[`Autoprefixer 1`] = `
":-webkit-full-screen a{
display:flex
}

:-moz-full-screen a{
display:flex
}

:-ms-fullscreen a{
display:-ms-flexbox;
display:flex
}

:fullscreen a{
display:-ms-flexbox;
display:flex
}"
`;

exports[`CSS-O (Optimizer) 1`] = `
".elem{
color:red
}"
`;

exports[`Calc Keep 1`] = `
".elem{
width:calc(100px + 2%)
}"
`;

exports[`Calc Trivial 1`] = `
".elem{
width:310px
}"
`;

exports[`Calc Variable 1`] = `
".elem{
width:210px
}"
`;

exports[`Clearfix 1`] = `
".row:after{
content:'';
display:block;
clear:both
}"
`;

exports[`Color Function 1`] = `
".yellow{
color:rgba(255,220,0,.9)
}"
`;

exports[`Config Maps 1`] = `
"h1{
background:darkyellow
}"
`;

exports[`Custom Media 1`] = `
"@media (max-width:30em){
body{
font-size:12px
}
}"
`;

exports[`Easing Gradients 1`] = `
".demo{
background:linear-gradient(#111 0%,rgba(17,17,17,.738) 19%,rgba(17,17,17,.541) 34%,rgba(17,17,17,.382) 47%,rgba(17,17,17,.278) 56.5%,rgba(17,17,17,.194) 65%,rgba(17,17,17,.126) 73%,rgba(17,17,17,.075) 80.2%,rgba(17,17,17,.042) 86.1%,rgba(17,17,17,.021) 91%,rgba(17,17,17,.008) 95.2%,rgba(17,17,17,.002) 98.2%,rgba(17,17,17,0) 100%)
}"
`;

exports[`Easings 1`] = `
".snake{
transition:all 600ms cubic-bezier(.47,0,.745,.715)
}"
`;

exports[`Flexbox Fix 1`] = `
".elem{
-ms-flex:1 1 0%;
flex:1 1 0%
}"
`;

exports[`Flow Root 1`] = `
".selector{
display:flow-root;
-webkit-column-count:1;
-moz-column-count:1;
column-count:1
}"
`;

exports[`Gradient Fix 1`] = `
".elem{
background-image:linear-gradient(green,rgba(0,128,0,0))
}"
`;

exports[`Grid KISS 1`] = `
".gridTest{
display:grid;
-ms-flex-line-pack:justify;
align-content:space-between;
grid-template:\\"a a\\" 120px \\"b c \\" 1fr \\"e e\\" 60px/30% 1fr
}

.gridTest>header{
grid-area:a;
-ms-flex-item-align:start;
align-self:start
}

.gridTest>.sidebar{
grid-area:b
}

.gridTest>main{
grid-area:c
}

.gridTest>footer{
grid-area:e;
justify-self:center;
-ms-flex-item-align:end;
align-self:end
}

@supports not (grid-template-areas:\\"test\\"){
.gridTest{
position:relative;
display:block;
width:100%;
height:100%
}

.gridTest>*{
position:absolute;
box-sizing:border-box
}

.gridTest>header{
max-height:120px;
width:100%;
top:0;
left:0
}

.gridTest>.sidebar,.gridTest>main{
height:calc(100% - 180px);
width:30%;
top:120px;
left:0
}

.gridTest>main{
width:70%;
left:30%
}

.gridTest>footer{
transform:translateX(-50%);
max-height:60px;
max-width:100%;
bottom:0;
left:50%
}
}

@media screen and (min-width:0\\\\0){
.gridTest{
position:relative;
display:block;
width:100%;
height:100%
}

.gridTest>*{
position:absolute;
box-sizing:border-box
}

.gridTest>header{
max-height:120px;
width:100%;
top:0;
left:0
}

.gridTest>.sidebar,.gridTest>main{
height:calc(100% - 180px);
width:30%;
top:120px;
left:0
}

.gridTest>main{
width:70%;
left:30%
}

.gridTest>footer{
transform:translateX(-50%);
max-height:60px;
max-width:100%;
bottom:0;
left:50%
}
}"
`;

exports[`Hex with Alpha 1`] = `
".blue{
color:rgba(17,17,255,.8)
}"
`;

exports[`Import Basic 1`] = `
".root{
color:red
}"
`;

exports[`Import with Asset URL 1`] = `
".image{
background:url(../fixtures/formula.png)
}"
`;

exports[`Import with Merge 1`] = `
".section{
color:#00d;
background:#333
}"
`;

exports[`Import with SVG URL 1`] = `
".preloader{
background:url(../fixtures/other/logo.svg);
border:1px solid #000
}"
`;

exports[`Import with Webfonts 1`] = `
"@font-face{
src:url(../fixtures/font/SourceSansPro-Light.otf.woff2) format(\\"woff2\\"),url(../fixtures/font/SourceSansPro-Light.otf.woff) format(\\"woff\\");
font-family:Source Sans Pro;
font-weight:300;
font-style:normal
}"
`;

exports[`Improved Color Palette 1`] = `
".red{
color:#ff4136
}"
`;

exports[`Input Style Fixes 1`] = `
"input[type=range]::-webkit-slider-runnable-track{
-webkit-appearance:none;
height:3px
}

input[type=range]::-moz-range-track{
-moz-appearance:none;
height:3px
}

input[type=range]::-ms-track{
height:3px
}

input[type=range]{
-webkit-appearance:none
}

input[type=range]::-webkit-slider-thumb{
-webkit-appearance:none;
width:16px;
height:8px
}

input[type=range]::-moz-range-thumb{
-moz-appearance:none;
width:16px;
height:8px
}

input[type=range]::-ms-thumb{
width:16px;
height:8px
}

input[type=range]::-moz-focus-outer{
border:0
}"
`;

exports[`Lost Grid 1`] = `
".grid{
width:calc(24.975% - 22.5px)
}

.grid:nth-child(n){
float:left;
margin-right:30px;
clear:none
}

.grid:last-child{
margin-right:0
}

.grid:nth-child(12n){
margin-right:0;
float:right
}

.grid:nth-child(12n+1){
clear:both
}"
`;

exports[`Magic Animations 1`] = `
".animation{
animation-name:magic
}

@keyframes magic{
0%{
opacity:1;
transform-origin:100% 200%;
transform:scale(1,1) rotate(0deg)
}

to{
opacity:0;
transform-origin:200% 500%;
transform:scale(0,0) rotate(270deg)
}
}"
`;

exports[`Media Query Min/Max 1`] = `
"@media screen and (min-width:500px) and (max-width:1200px){
.elem{
display:block
}
}"
`;

exports[`Nested Basic 1`] = `
"body h1{
font-weight:700
}

body h2{
font-weight:400
}"
`;

exports[`Nested Parent Selector 1`] = `
"body ul li:first-child{
margin-top:0
}"
`;

exports[`Normalize.css 1`] = `
".before{
color:#ff4136
}

html{
line-height:1.15;
-ms-text-size-adjust:100%;
-webkit-text-size-adjust:100%
}

h1{
font-size:2em;
margin:.67em 0
}

hr{
box-sizing:content-box;
height:0;
overflow:visible
}

details,main{
display:block
}

code,kbd,pre,samp{
font-family:monospace,monospace;
font-size:1em
}

a{
background-color:transparent;
-webkit-text-decoration-skip:objects
}

abbr[title]{
-webkit-text-decoration:underline dotted;
text-decoration:underline dotted
}

b,strong{
font-weight:bolder
}

small{
font-size:80%
}

img{
border-style:none
}

svg:not(:root){
overflow:hidden
}

button,input,optgroup,select,textarea{
margin:0
}

button{
overflow:visible;
text-transform:none
}

[type=button],[type=reset],[type=submit],button{
-webkit-appearance:button
}

[type=button]::-moz-focus-inner,[type=reset]::-moz-focus-inner,[type=submit]::-moz-focus-inner,button::-moz-focus-inner{
border-style:none;
padding:0
}

[type=button]:-moz-focusring,[type=reset]:-moz-focusring,[type=submit]:-moz-focusring,button:-moz-focusring{
outline:1px dotted ButtonText
}

fieldset{
padding:.35em .75em .625em
}

input{
overflow:visible
}

legend{
color:inherit;
display:table;
max-width:100%;
white-space:normal
}

progress{
display:inline-block;
vertical-align:baseline
}

select{
text-transform:none
}

textarea{
overflow:auto
}

[type=checkbox],[type=radio],legend{
box-sizing:border-box;
padding:0
}

[type=number]::-webkit-inner-spin-button,[type=number]::-webkit-outer-spin-button{
height:auto
}

[type=search]{
-webkit-appearance:textfield;
outline-offset:-2px
}

[type=search]::-webkit-search-cancel-button,[type=search]::-webkit-search-decoration{
-webkit-appearance:none
}

::-webkit-file-upload-button{
-webkit-appearance:button;
font:inherit
}

summary{
display:list-item
}

[hidden],template{
display:none
}

.after{
color:#2ecc40
}"
`;

exports[`Pleeease Filters 1`] = `
".box{
filter:url('data:image/svg+xml;
charset=utf-8,<svg xmlns=\\"http://www.w3.org/2000/svg\\"><filter id=\\"filter\\"><feGaussianBlur in=\\"SourceAlpha\\" stdDeviation=\\"20\\" /><feOffset dx=\\"17\\" dy=\\"17\\" result=\\"offsetblur\\" /><feFlood flood-color=\\"rgba(0,0,255,1)\\" /><feComposite in2=\\"offsetblur\\" operator=\\"in\\" /><feMerge><feMergeNode /><feMergeNode in=\\"SourceGraphic\\" /></feMerge></filter></svg>#filter');
-webkit-filter:drop-shadow(16px 16px 20px #00f);
filter:drop-shadow(16px 16px 20px #00f)
}"
`;

exports[`Pseudoelements 1`] = `
"a:before{
color:#ff4136
}"
`;

exports[`RGBA with Hex 1`] = `
".red{
color:rgba(255,17,17,.8)
}"
`;

exports[`Responsive Type 1`] = `
"html{
font-size:calc(7.60465px + 1.04651vw)
}

@media screen and (min-width:1280px){
html{
font-size:21px
}
}

@media screen and (max-width:420px){
html{
font-size:12px
}
}"
`;

exports[`Sassy Mixins 1`] = `
"h1{
color:#ff4136
}"
`;

exports[`Sassy Variables 1`] = `
"h1{
background:#ff4136
}"
`;

exports[`Selector matches() 1`] = `
"p.special,p:first-child{
color:#ff4136
}"
`;

exports[`System UI 1`] = `
"body{
font-family:system-ui,-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Oxygen,Ubuntu,Cantarell,Fira Sans,Droid Sans,Helvetica Neue
}"
`;

exports[`Transform Shortcut 1`] = `
".transform{
transform:scale3d(2,1,1) translate3d(10px,20px,0)
}"
`;

exports[`Will Change Compat 1`] = `
".scaled{
-webkit-backface-visibility:hidden;
backface-visibility:hidden;
will-change:width
}"
`;

exports[`zIndex 1`] = `
".first{
z-index:1
}

.second{
z-index:2
}"
`;
